From 00d3cfbde2134ca20e2ea0fe00b2be0168ea4b02 Mon Sep 17 00:00:00 2001
From: yoko2001 <1766649097@qq.com>
Date: Mon, 28 Feb 2022 00:29:03 +0800
Subject: [PATCH 5/5] bug fix3

---
 include/linux/swap.h |  1 +
 mm/swapfile.c        | 12 +++++++-----
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/include/linux/swap.h b/include/linux/swap.h
index 0244d587..85406048 100644
--- a/include/linux/swap.h
+++ b/include/linux/swap.h
@@ -444,6 +444,7 @@ extern bool reuse_swap_page(struct page *, int *);
 extern int try_to_free_swap(struct page *);
 struct backing_dev_info;
 extern int init_swap_address_space(unsigned int type, unsigned long nr_pages);
+extern int init_hpswap_list(struct swap_info_struct *sis);
 extern void exit_swap_address_space(unsigned int type);
 
 extern int get_swap_slots(int n, swp_entry_t *slots);
diff --git a/mm/swapfile.c b/mm/swapfile.c
index e90517c9..a8d5adc2 100644
--- a/mm/swapfile.c
+++ b/mm/swapfile.c
@@ -766,7 +766,9 @@ no_page:
 	si->flags -= SWP_SCANNING;
 	return n_ret;
 }
-
+static bool hpnode_list_empty(struct swap_hp_list shl){
+	return (shl.head.next) == (&shl.tail);
+}
 static int scan_hpswap_map_slots(struct swap_info_struct *si,
 			       unsigned char usage, int nr,
 			       swp_entry_t slots[])
@@ -775,7 +777,7 @@ static int scan_hpswap_map_slots(struct swap_info_struct *si,
 	unsigned long last_in_cluster = 0;
 	int latency_ration = LATENCY_LIMIT;
 	int n_ret = 0;
-	if 
+
 	if (nr > SWAP_BATCH)
 		nr = SWAP_BATCH;
 
@@ -827,11 +829,11 @@ checks:
 	//we are going to check every hugepages, and found the first free one, get its offset and so on
 	struct swap_hp_node *cur, *nxt;
 	cur = &si->hugepages.head;
-	nxt = list_entry(&cur->list, struct swap_hp_node, list);
+	nxt = list_entry(cur->list.nxt, struct swap_hp_node, list);
 	while(nxt != &si->hugepages.tail){
 		// move backward by one
-		cur = next;
-		next = list_entry(&cur->list, struct swap_hp_node, list);
+		cur = nxt;
+		nxt = list_entry(cur->list.next, struct swap_hp_node, list);
 
 		//this page has been used, we check the next one
 		if (cur->flags & HPSWP_GOOD && cur->flags & HPSWP_USED)
-- 
2.33.1.windows.1

